sklearn/utils/_array_api.py

```diff
@@ -463,14 +478,14 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
     if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+        sample_score = xp.astype(
+            sample_score, max_precision_float_dtype(xp, device(sample_score))
```

Comment:
Not being able to write `xp.float64` makes things hardened to develop for. But I see that it's necessary.

In the end, we'll need to forbid using directly `xp.float64` because it is not portable.